### DISCUSSION TITLE
New approach for allowing games to perform own selection logic

### DISCUSSION THREAD

**Discussion by badosu** [beyond-all-reason/RecoilEngine#668] [4. Feature Requests and Ideas]
I noticed [this](https://github.com/beyond-all-reason/spring/issues/288) was implemented for allowing games to take over box selection from engine.

I went through all current Recoil consumers and was able to find none that makes use of Set/GetSelectionBoxHandledByEngine.

At the same time, I believe this is an important ability for the engine to support (games that don't want to use engine defaults/want to implement own selection logic), and have delved into and dealt with smart select (one such widget for own selection logic, used in BAR) issues and their causes: mostly inability to infer selection events due to the complexity managing mousepress/release owner _and_ having to ignore engine triggered selections.

I have a [WIP PR](https://github.com/beyond-all-reason/spring/pull/667) that implements a `BoxSelection(anchorx, anchory, pivotx, pivoty)` callin that allows games to:

- Listen for selection decoration/effects
- Use event for some special logic - remember mouserelease only is listened to if we return true on mousepress hence invalidating engine selection
- Perform our own selection logic with passed mouse coordinates and return true to invalidate engine selection

[PS: As well observed by @sprunk it would also make sense to expose `BoxSelection{Move,Update}` for dynamic selection boxes (logic without releasing box)]

Apart from that it also addresses a few shortcomings I identified with the previous approach:
- Minimap box selection wasn't being considered
- Hardcoded ctrl/shift for remove-from/append-to selection wasn't being ignored when (and only when) in box selection and box selection not handled by engine
- Instead of just returning early, still performing individual unit selection if in box selection and box selection not handled by engine

Since this seems (to me) to be a more appropriate model for exposing this ability to games I suggest allowing this callin to be available. I am unsure if `{Get,Set}BoxSelectionHandledByEngine` still have any relevant role in this circumstance.

Nothing changes for consumers that don't intend to perform own selection logic and the incurred overhead is minimal (checking that boxSelectionList.size == 0 on mouserelease without mouseowner and dragselectionthreshold met).

Full selection box externalization would perhaps include things like `DrawBoxSelection` and/or other control callouts, but this does not seem to be a demanded feature (at least that I know).

PS: This is already somewhat possible, at least Zero-K and BAR have special logic for selections that in some way or another override the default engine one. However they rely on 1) checking mouse state with `Update` which: is not exactly reliable nor desirable to be used, and at least for BAR is a source of issues and unnecessarily complex code; and 2) nooping or double selecting to ignore engine default.

**Comment by badosu** [comment]
cc @MaDDoXbr @GoogleFrog @Ruwetuin @springraaar @ForbodingAngel @skymyj

**Reply by GoogleFrog** [reply]
My main concern is that of reinventing the wheel. Of making a parallel implementation of something that already works, adding to complexity and maintenance overhead.

I am very happy with the selection filtering I implemented for ZK. It doesn't feel janky or incomplete to me, so from the outset I don't see what engine support could add. Selection doesn't seem expensive enough to warrant optimisation via engine support. So I have to look to your post for the motivation behind a selection system, but in it I see the motivation coming from the shortcomings of how BAR implements it, rather than something inherent to the engine.

The ZK widget basically just uses widget:SelectionChanged(selectedUnits, subselection). Returning a list of units from this callin updates the selection seamlessly, the change is intercepted and applied in the widget handler. There is no use of widget:Update, and the only interaction of the widget with the mouse is with widget:MousePress, which just contains a timer to check for double clicks. The widget works how I want it to and the API with the widget handler is useful and clean. It handles selection on the minimap and control groups just fine because it trusts the list of selected units sent to it by the engine. There is no concept of a minimap within the widget, and it's only concept of control groups is to disable filtering when a key mapped to a control group is held (because control group recall shouldn't be filtered).

I haven't used the BAR widget, but I read it just now, and I can see the source of your problems. Rather than return a new list of selected units, the BAR widget sets a flag in WG then waits for the next widget:Update to handle the selection event. Update has 200 lines of code that I haven't looked at closely, but it seems to read mouse coordinates, then using coordinated it saved via.... WG.SmartSelect_MousePress2 (that is called directly from the widget handler (?!?)) it defines a screen-rectangle, asks spring for the units in that screen-rectangle, and does some filtering on them. This does seem pretty bad and I imagine deferring the handling until Update would cause unpolished jank. I don't know what set of requirements led to the widget being written like this. I noticed that the BAR widget was written in 2011 while I wrote mine from scratch in 2017. I don't think there was a relevant new feature in that time, but it isn't unusual for widgets from around 2010 to need a rewrite.

ZK and BAR seem to have the same functionality in their widget handler to allow this, so if ZK edited in this behaviour, then BAR copied the changes, so perhaps its worth putting this functionality in the default widget handler. As far as I can tell, widget:SelectionChanged is entirely made up by luaUI anyway. So I think the first step would be to rewrite the BAR widget rather than change the engine, and only look into an engine change if making a nice selection filtering widget with the BAR requirements turns out to be impossible.

**Reply by GoogleFrog** [reply]
That said, there is some room for improving selection in the engine. I just think the approach should be an extension of the existing SelectionChanged callin rather than the proposed addition of a special form of MouseReleased. Sending the new list of selected units to widgets for filtering seems much neater than having the widgets generate a list via feeding some coordinates to a rectangle. A small filter is nicer than being notified that a rectangle happened, then having to decide whether to completely reimplement engine behaviour or let the engine do its thing.

Anyway, I think an engine implementation of SelectionChanged could contain information about where the change was from. Was it from a control group, a dragged box on the main screen, a dragged box on the minimap, a unit dying, a selectkey, or from somewhere in lua? This would mean my widget would stop having to guess whether a change was caused by a control group hotkey or double click.

I don't anticipate getting any use out of `BoxSelection(anchorx, anchory, pivotx, pivoty)` (well, for its intended purpose. I could use it to hax a way to get MouseReleased events that lua "shouldn't" be getting) because the ZK selection widget already handles selection in a neater way, and it handles it for other ways of selecting units such as selectkeys and minimap boxes.

**Reply by sprunk** [reply]
>  [in ZK] there is no use of widget:Update, and the only interaction of the widget with the mouse is with widget:MousePress
>
> I don't know what set of requirements led to the [BAR] widget being written like this. 

Units are actually getting selected as you drag in BAR, instead of just highlighted like in ZK. This sounds like it would require doing stuff in Update because there is no SelectionChanged being sent continuously.

**Reply by badosu** [reply]
Yeah, it's not that different from ZK except BAR does dynamically.

ZK requirements are more lax because you just need to know that you moved from mousepressed to mousereleased, but it's still not a strictly proper implementation. But it's good enough as long as your events do not happen between Updates.

**Reply by badosu** [reply]
As long as I agree that in general code quality in BAR widgets is not amazing compared to ZK it is a bit annoying that this is interpreted as a engine fix for a BAR ad-hoc problem.

The requirements for smartselect are different from the ones I observed in ZK as well, as highlighted by sprunk above.

I will read the comments with more tranquility over the few days but I'll anticipate that this _will enable_ games to externalize all their selection logic feature paired with engine. And keep in mind that any attempt at an implementation in lua, be it in production or not is a workaround, with its own flaws due to not being correct or deterministic.

**Reply by sprunk** [reply]
Wait, how are the current approaches not correct or not deterministic, or don't enable games to externalize logic? Can you give an example of incorrect behaviour?

In ZK's `SelectionChanged` approach, it doesn't matter whether the dimensions of the box as last seen in `Update` are different compared to the actual dimensions as they were at mouse release, or even what the box dimensions were at release all. This is because `SelectionChanged` only receives the set of units that the engine (correctly and deterministically!) decided were inside the box at the time of mouse release, and not any information about the box itself.

BAR's approach has two parts - one is the mouse drag, where correctness/deterministicness is nebulous to define because it's a continuous motion but also where these two don't really matter because the mouse release has the final say, and the second part is the mouse release part where `SelectionChanged` already largely enables a game to externalize logic and do everything it needs. The main thing which is lacking is to let games distinguish whether `SelectionChanged` came from a box-drag or some other source (like pressing a hotkey for a control group), but see below https://github.com/beyond-all-reason/spring/discussions/668#discussioncomment-5199215 . Is there anything that `BoxSelectionRelease(...)` allows to do with selection that `SelectionChanged(..., source = "box-drag")` wouldn't?

**Reply by badosu** [reply]
For ZK, whatever relies on `holdingForSelection` will be non deterministic and incorrect. Anything that depends on reimplementing mouserelease without owner basically.

**Reply by badosu** [reply]
For smartselect/BAR, it will still be incorrect with (and with only) proper `SelectionChanged` due to inability to rely on engine passed selection caused by hardcoded engine modifiers.

**Reply by sprunk** [reply]
>  whatever relies on `holdingForSelection` will be non deterministic and incorrect.

Incorrect how? Is this just for the case where a thing is not considered as being in the box in the 0.1th of a second between Update calls?
Nondeterministic how? Is it just because Update can sometimes fire after 0.2s instead of 0.1?

This kind of incorrect/nondeterministic doesn't matter, it's on timescales that people don't notice anyway, and under that mindset almost everything that uses discrete events for continuous things (in particular most uses for `Update`, but also for example `MouseMove` which doesn't fire an event for every single pixel it moves over) would be nondeterministic and incorrect.

**Reply by sprunk** [reply]
> inability to rely on engine passed selection caused by hardcoded engine modifiers.

Good point. Sounds like a motivation to allow unbinding (or rebinding, etc) selection modkeys like you did with camera modkeys recently.

**Reply by sprunk** [reply]
To summarize https://github.com/beyond-all-reason/spring/discussions/668#discussioncomment-5199424

Even though I think the claims about incorrect/nondeterministic dragging don't really matter, I still like the "on selection box dragging" event because it would allow reducing some ugly boilerplate.

You didn't seem to claim that SelectionChanged is nondeterministic, and only claimed it's incorrect due to modifiers, which can be addressed, so I think I dislike the "on selection box release" event for being inferior to an improved version of SelectionChanged that supplies the source of the change.

**Reply by badosu** [reply]
> > inability to rely on engine passed selection caused by hardcoded engine modifiers.
> 
> Good point. Sounds like a motivation to allow unbinding (or rebinding, etc) selection modkeys like you did with camera modkeys recently.

Yep, that's the idea. This was a step on the roadmap for that: https://github.com/beyond-all-reason/spring/pull/562

Unfortunately I hit a (perceived or not) impasse and don't feel comfortable proceeding without additional input.

**Reply by badosu** [reply]
> Even though I think the claims about incorrect/nondeterministic dragging don't really matter

I agree, for most cases its 'good enough'. But good enough is not enough when it could be correct, as in the engine providing a sane and sensible API for handling stuff that is worked around against it instead.

Non-deterministic as in the internal lua state mapped to reflect what would be internal engine state depends on non deterministic factors, such as user input. Although for the crushing majority of cases it is mitigated by a simpler API, (such as just mapping mousepress,mouserelease on update).

> You didn't seem to claim that SelectionChanged is nondeterministic, and only claimed it's incorrect due to modifiers, which can be addressed, so I think I dislike the "on selection box release" event for being inferior to an improved version of SelectionChanged that supplies the source of the change.

Exactly, although more important than the source of the change (which is a nice addition but in tandem with the mantra of working around engine deficiencies) is not depending on an API that was devised for a different time and purpose and that is also worked around to provide the functionality it should without 'hacks'. (case in point selectionchanged listening to a callin that sends exact coordinates for buttons and pages for an ad-hoc and engine coupled lua ui)

**Reply by badosu** [reply]
Just for the record, I use those words to denote engine deficiencies that reflect on game code and I might be more precious than might be normally accepted for everyday usage. This is mostly because I have an opinion that the API exposed by the engine should help game developers instead of being a hindrance to them, and most importantly enabling games to consume state lifecycles that avoid requiring an inconsistent or difficult to maintain (feature parity wise) reflection of internal engine state.

**Reply by sprunk** [reply]
> Non-deterministic as in the internal lua state mapped to reflect what would be internal engine state depends on non deterministic factors, such as user input.

Okay, but how is that bad? Doesn't that apply to any user input? Wouldn't the proposed dragged box callout ultimately also just reflect user input? I can't think of any use that would require input to be deterministic to do something useful with it, you just take the unpredictable input "unquestioningly" and do something about it. Say a user is dragging a box from (0, 0) to (x, y). It's not possible to predict x and y because they are user input, but you generally just pass them to a `Spring.GetUnitsInScreenRectangle(x, y)` function regardless.

**Reply by badosu** [reply]
> Doesn't that apply to any user input?

No, if you perform a selection fast enough you might get inconsistent internal state. For simpler models such as mapping mousepress/release this is not the biggest issue, but with higher cpu constraints it might get exacerbated. If you ensure a proper select box callin is called then you get a correct response, as long as SDL captured the events necessary.

**Reply by badosu** [reply]
> I can't think of any use that would require input to be deterministic to do something useful with it

Sure but the question is about having to work around the engine on a faulty model that might not suffice for all games and/or requirements. It's not a huge effort to provide a proper API.

**Reply by badosu** [reply]
Just an example, regardless of initial proposal `BoxSelection` or proper `SelectionChanged` with filterable input you can get a correct state for the rectangle you want to select. This is relevant for smartselect that works dynamically and continuously maps the selection box rectangle, at the end of the selection you want to 'lock' the proper geometry of the selection. Otherwise fast selections don't do their job.

Keep in mind in this case we don't want or trust the engine defined selection, but our own logic.

**Comment by sprunk** [comment]
Loose thoughts:
 * having a way to distinguish the dragging and the release would be good, but having just the on-release version is already an improvement if you're low on mana to implement both.
 * deprecating `{S,G}etBoxSelectionHandledByEngine` sounds good if you can return a bool from the on-release version of the new box callin.
 * handling selection on minimap sounds good but ideally there would also be a way to distinguish it from the main view world selection (since it's easy to click the minimap by accident when dragging diagonally and get a completely different selection than intended).
 * `DrawBoxSelection` sounds good but very low priority, the regular plain box is fine.
 * it would be good if this was fit for general cases where you might want to drag a box, not just selection. An example use story would be, if you hold some button and drag it displays some information about the inside of the box (for example total value of mexes) without selecting. As far as I can tell your proposed interface already makes this possible, and there may be better ways to do this (area commands?), but it might be worth keeping that use case in mind for design.

**Reply by badosu** [reply]
[Removed my previous main thread comments to keep the thread organized]

I agree with the points, that sounds good.

With regard to:

> it would be good if this was fit for general cases where you might want to drag a box, not just selection. An example use story would be, if you hold some button and drag it displays some information about the inside of the box (for example total value of mexes) without selecting. As far as I can tell your proposed interface already makes this possible, and there may be better ways to do this (area commands?), but it might be worth keeping that use case in mind for design.

Unfortunately it doesn't since by issuing a command that relies on the mousepress (e.g. fight command) - or capturing mouse ownership via lua mousepress - an activeReceiver is registered, precluding engine from reaching the codepath for boxselection (the callin will never be called).

In this case some kind of way for the game to signal it wants to activate a custom boxselection must be devised, e.g. `Spring.StartBoxSelection(x[opt=mousex], y[opt=mousey])`. This callout can either enqueue a custom selection for next mousepress + dragthreshold ignoring/before any activereceivers (this might require a companion `Spring.CancelBoxSelection`) or require to be invoked during mousepress itself (regardless of mouseownership).

[Edit: An alternative is having the callin called even _before_ checking for activeReceiver, in which circumstance its presence can be signalled as an argument. Then some logic can be devised for custom box and return values but it sounds a bit convoluted and dangerous.]

With regard to:

> handling selection on minimap sounds good but ideally there would also be a way to distinguish it from the main view world selection (since it's easy to click the minimap by accident when dragging diagonally and get a completely different selection than intended)

This can be decided by the game on how to handle on the current approach already. In this situation the game can check if `Spring.IsAboveMiniMap(anchorx|pivotx, anchory|pivoty)` and decide if it wants to handle a custom logic for minimap selection, ignore it and perform selection underneath, invalidate selection of any kind, etc...

**Reply by badosu** [reply]
I am up for extensions of the initially modest proposal, however I think it would only be worth implementing it if a consumer signalled it is currently something they'd use.

I know for example that BAR would make use of `BoxSelection{Move,Update}` at least. The first iteration would be to ensure at least `BoxSelection{Release,}` is available for correctness and most basic support.

**Reply by sprunk** [reply]
>  I think it would only be worth implementing it if a consumer signalled it is currently something they'd use.

For the on-release callin: I think ZK probably wouldn't immediately use the currently implemented event for actual selection-related stuff, because we already have a robust implementation via existing means so would gain essentially nothing from switching. I think ZK, or at least I personally as a widget maker, might still use it but and mostly to get box dragging for free for a different fairly low-priority feature.

For the on-dragging callin: I don't think ZK would use it currently but I also don't see why we wouldn't use it later. Recently our players wanted us to step up the hype game, and the "select as you drag" feature apparently make a game look more modern, so maybe a player would want to port it, in which case I'd prefer him to use the callin if it was available.

> Unfortunately it doesn't since by issuing a command that relies on the mousepress (e.g. fight command) - or capturing mouse ownership via lua mousepress - an activeReceiver is registered, precluding engine from reaching the codepath for boxselection (the callin will never be called).

I didn't necessarily mean that area commands would be used alongside the selection, rather they are an alternative. Consider this:
```lua
local totalMetalInBox
function widget:BoxSelectionDrag(x1, y1, x2, y2)
  if Spring.Utilities.IsKeyPressed('E') then -- I don't actually know the key interfaces, pretend this checks if the key is pressed
    totalMetalInBox = 0
    local features = Spring.GetFeaturesInScreenRectangle(x1, y1, x2, y2) -- actually only exists for Units, feature request?
    for i = 1, #features do
      local metal = Spring.GetFeatureResources(features[i])
      totalMetalInBox = totalMetalInBox + metal
    end
  end
end

function widget:BoxSelectionRelease()
  if Spring.Utilities.IsKeyPressed('E') then
    totalMetalInBox = nil
    return false -- we only came here for the reclaim display, don't select
  else
    return true
  end
end

function widget:DrawWorld()
  if Spring.Utilities.IsKeyPressed('E') and totalMetalInBox then
    Spring.Utilities.DrawTextAtCursor("Metal: " .. totalMetalInBox)
  end
end
```
Now you can hold E and drag a box to see how much metal is in the box, and releasing the mouse button won't touch your current selection. A similar thing can already be achieved with the area reclaim command - you also press E and drag (and there's some widget that calculates the total in the area if the currently selected command is reclaim, similar to the one above). The motivation to use box drag instead of the area command here is that
 * you don't need a unit with the Reclaim command available to be selected to use that interface. More generally, you don't need a unit selected at all. In theory you can make commands available globally as well, but that has different trade-offs (it could be confusing since you aren't actually giving the command to anything and it's not obvious it's just for the display UI).
 * different shape (not just circle vs rectangle, but also screenspace vs worldspace coordinates - both affect precision and intuitiveness)
 * different control (hold button vs press button once then press Escape to cancel)
 * possibly others, I'm just going with the example that first came to my head but there may be more realistic ones.

**Reply by badosu** [reply]
Yes, this would be feasible already assuming no activeReceiver is present (e.g. your widget does its own internal logic without relying on commands or capturing mouse ownership).

**Reply by GoogleFrog** [reply]
> For the on-dragging callin: I don't think ZK would use it currently but I also don't see why we wouldn't use it later. Recently our players wanted us to step up the hype game, and the "select as you drag" feature apparently make a game look more modern, so maybe a player would want to port it, in which case I'd prefer him to use the callin if it was available.

We already have a system for this. Dragging a selection box in ZK already draws a highlight around the units that will be selected. See https://github.com/ZeroK-RTS/Zero-K/blob/master/LuaUI/Widgets/api_preselection.lua

Having `widget:BoxSelectionDrag(x1, y1, x2, y2)` wouldn't be bad per-se, but it seems a bit redundant when lua seem to be able to handle it fine. The engine has a bit of a stability advantage and performance could go either way, but probably isn't worth worrying about. A calling which sent the list of units which would be selected, rather than the coordinates, would be neater but still not particularly significant.

> function widget:BoxSelectionRelease()
  if Spring.Utilities.IsKeyPressed('E') then
    totalMetalInBox = nil
    return false -- we only came here for the reclaim display, don't select
  else
    return true
  end
end

This seems pretty contrived to me. It is easy to write a widget to return true on MousePressed, record the coordinate, draw a box while the mouse is held, get whatever it wants from the drawn rectangle, then do some sort of aggregation on them. This is exactly the sort of thing luaUI is good at. There is no reason to intermingle the logic of the widget with the inbuilt selection system. All the required mouse events exist and work well, and the widget is more flexible. Eg what if the widget wants to drag edges of a circle, or draw a different sort of selection box.

I don't think we should be adding new ways to do things that lua already seems to do very well, at least until someone tries to implement the feature in lua and hits a stumbling block. Unused bonus features are likely to silently break at some point and contribute to code complexity. I'm all for features, even without clear uses, if it would let a dev do something (possibly with new lua features) that lua either couldn't do or could only do with a bunch of hax.

**Reply by sprunk** [reply]
> It is easy to write a widget to return true on MousePressed, record the coordinate, draw a box while the mouse is held, get whatever it wants from the drawn rectangle, then do some sort of aggregation on them. This is exactly the sort of thing luaUI is good at.

I think the engine should, in general, provide basic building blocks for stuff regardless of how easy it is to implement them in Lua. This is because it lowers entry barrier for new gamedevs (the engine is centralized; if the interface exists in an obscure widget spread across games they will be unaware of its existence), prevents mistakes (consider that the BAR version of preselection is terrible) and generally establishes a common standard (easier to work if the fundamental blocks are the same everywhere). I consider dragging a rectangle a basic building block.

> Eg what if the widget wants to drag edges of a circle, or draw a different sort of selection box.

I think dragging a rectangle is a basic building block in a way that dragging a circle isn't. It's fine to make the user implement circle dragging on his own.

> There is no reason to intermingle the logic of the widget with the inbuilt selection system. 

Dragging the mouse is fundamentally an act of drawing a rectangle with no particular meaning. The inbuilt selection system is that selecting units just happens to be the default meaning assigned to that rectangle. It only looks like dragging the rectangle is an integral part of the inbuilt selection system because it's not possible to give any other meaning to that rectangle, but I would like to expose to Lua a way to give the rectangle a different meaning than selection.

> I'm all for features, even without clear uses, if it would let a dev do something (possibly with new lua features) that lua either couldn't do or could only do with a bunch of hax.

I think we agree but have different thresholds of what is hax.

> Unused bonus features are likely to silently break at some point and contribute to code complexity.

Generally I agree but in this case it's an engine dev asking for feedback for an existing PoC that is expected to see use.

**Reply by badosu** [reply]
Anything currently in Lua that performs selection logic is hacky already, it might be good enough but its neither efficient, correct or deterministic. One either puts together their own Lua representation of Engine select state + logic in a constant loop, or a diminished version of it, using a deficient API.

The issue is a fundamental one, not tied to necessarily a certain games requirements (even if the motivation to fix it is due to it), regardless of how good you think your code is.

At a more constructive level, this enables games to actually implement a sane pipeline for their own selection box logic, which is something you'd want your engine to be able to support anyway (even if no consumer currently does it properly).

**Reply by badosu** [reply]
Heck, even `SelectionChanged` is a hack on top of a Callin that is not proper for that.

**Reply by badosu** [reply]
In fact, I should make a priority for providing a proper `SelectionChanged` even before this, since all consumers use the current hack in a very inefficient pipeline. `LayoutButtons` performs logic that is unnecessary for most games and still requires games to invoke callouts to get the data that the proper callin should provide in the first place.

**Reply by sprunk** [reply]
Native `SelectionChanged` sounds good, and a good opportunity to add a "source of change" parameter to it? "box-drag", "control-group", "point-click", "death", "lua", ...
﻿
In fact, `SelectionChanged(..., source = "box-drag")` sounds like it would obsolete the proposed event to finalize the dragging of a box selection (and be somewhat simpler since you get to skip the `Spring.GetUnitsInScreenRectangle` call).

**Reply by badosu** [reply]
For dynamic selection boxes plugins it would suffice for the definition of the selection and not needing to override the engine passed selection. yes

**Reply by GoogleFrog** [reply]
Native `SelectionChanged` sounds best. With sources (engine tends to send integers as sources and put the meaning in a table of constants). I am looking forward to `BoxSelection` too, but because it will make some hax a bit cleaner, not for legitimate use. `BoxSelectionUpdate` should be generally good for the preselection style highlighting or selection widgets though.

**Comment by GoogleFrog** [comment]
https://github.com/beyond-all-reason/spring/issues/567 Might be related.

**Reply by sprunk** [reply]
#567 is for the actual selection (regardless if you use the built-in engine box or some sort of Lua selection) and #668 is just for dragging the box (regardless if you use it to select units or anything else).

**Comment by GoogleFrog** [comment]
`BoxSelection(anchorx, anchory, pivotx, pivoty)` reminded me of a related issue with selection where having an event for "a selection box just happened" would be useful, but it would only make a hacky widget very slightly less hacky. I made a ticket about it https://github.com/beyond-all-reason/spring/issues/671

**Comment by badosu** [comment]
A per discussion with @sprunk [here](https://github.com/beyond-all-reason/spring/discussions/668#discussioncomment-5199215) I'm prioritizing a proper `SelectionChanged` callin (in place of `LayoutButtons`) instead of the initial proposal.

We might be able to rework this discussion to reflect future requirements of games for full externalization of selection, but since this is not a demand any consumer apparently has for now I won't be proceeding with it.

Thanks for the feedback and helping find a better solution.
